### PR TITLE
Fetch, Parse, and Create Documents for Statically Hosted Files

### LINKED ISSUE #2110: [FEAT]: Allow bulk web scraper to download files that it discovers
### What would you like to see?

So using the web scraper to browse a supplier's site. Site links to many PDF datasheets. It would be helpful if the scraper could download these to embed.

Tunable parameters would be helpfula s well.

E.g. allowed/excluded file types: pdf, docx etc
Maximum file size: (to prevent the scraper from downloading say a 300mb PDF or something)

### PR DESCRIPTION
**Pull request by angelplusultra**
### Pull Request Type

- [x] ✨ feat
- [ ] 🐛 fix
- [ ] ♻️ refactor
- [ ] 💄 style
- [ ] 🔨 chore
- [ ] 📝 docs

### Relevant Issues

resolves #2110 

### What is in this change?

This PR introduces an enhancement to the document upload modal: the web scraping input field can now fetch, parse, and convert statically hosted files (e.g., https://example.com/a-file.pdf) and API endpoints returning JSON (e.g., https://jsonplaceholder.typicode.com/posts) into application documents.

Currently Supporting:
### Currently Supporting:

- **Text**
  - Plain text: `.txt`, `.md`, `.org`, `.adoc`, `.rst`
  - HTML: `.html`
  - CSV: `.csv`
  - JSON: `.json`

- **Documents**
  - Word (DOCX): `.docx` (NOTE: `.doc` files are currently **not** supported, only `.docx`)
  - OpenDocument Text: `.odt`
  - PDF: `.pdf`
  - EPUB: `.epub`

- **Presentations**
  - PowerPoint (PPTX): `.pptx`
  - OpenDocument Presentation: `.odp`

- **Spreadsheets**
  - Excel (XLSX): `.xlsx`

- **Email**
  - MBOX: `.mbox`

- **Audio**
  - WAV: `.wav`
  - MP3: `.mp3`

- **Video**
  - MP4: `.mp4`
  - MPEG: `.mpeg`

- **Images**
  - PNG: `.png`
  - JPEG/JPG: `.jpg`

### Additional Information

The core functionality hinges on the Content-Type header of the response. If the mimetype is supported, your file will be pulled and processed.

### Developer Validations

- [x] I ran `yarn lint` from the root of the repo & committed changes
- [x] Relevant documentation has been updated
- [x] I have tested my code functionality
- [ ] Docker build succeeds locally

### CODE REVIEW

File: collector/processLink/convert/generic.js [hunk 1]
```diff
@@ -18,58 +23,104 @@ const { default: slugify } = require("slugify");
  */
 async function scrapeGenericUrl({
   link,
-  captureAs = "text",
```
Comment (timothycarambat) [inline collector/processLink/convert/generic.js:22]: The removal of `captureAs` breaks the collector here in the agent web-scrape and agent-flow webscraper. Since this function will also now _always_ generate a document.

 However, it seems that this is supposed to be replaced by the `processAsDocument` key.
 
 These two flags do two separate functions and can work together, though. `captureAs` basically determines what type of [content will we scrape from a site](https://github.com/Mintplex-Labs/anything-llm/blob/3c61e400d0f579b543246903baef532d0e0d5d57/collector/processLink/convert/generic.js#L170-L171)
 
 However,  because `captureAs` was removed as an option.arg we can no longer use it and `getPageContent` will always only get text.
 
 [`getLinkText`](https://github.com/Mintplex-Labs/anything-llm/blob/3c61e400d0f579b543246903baef532d0e0d5d57/collector/processLink/index.js#L33) still uses this arg so it would just be dropped

File: collector/processLink/convert/generic.js [hunk 2]
```diff
@@ -18,58 +23,104 @@ const { default: slugify } = require("slugify");
...
   processAsDocument = true,
   scraperHeaders = {},
   metadata = {},
 }) {
-  console.log(`-- Working URL ${link} => (${captureAs}) --`);
```
Comment (timothycarambat) [inline collector/processLink/convert/generic.js:27]: This log is actually very useful for debugging, observability, and when someone opens a GH issue "why did my website not get scraped" and it allows us to easily repro.

It is also just nice to have just to know the collector is working via logs.

File: collector/processLink/convert/generic.js [hunk 3]
```diff
@@ -18,58 +23,104 @@ const { default: slugify } = require("slugify");
...
-    chunkSource: `link://${link}`,
-    published: new Date().toLocaleString(),
-    wordCount: content.split(" ").length,
-    pageContent: content,
-    token_count_estimate: tokenizeString(content),
+  // If the content type is not supported, return an error
+  return {
+    success: false,
+    reason: `Unsupported content type: ${contentType}`,
+    documents: [],
```
Comment (timothycarambat) [inline collector/processLink/convert/generic.js]: Since the entire logic is now in a big `if` it would make more sense to check this as a guard clause at the top of the function so one big `if` is not wrapping the core of the function.

eg: 
```js
if(!(contentType in ACCEPTED_MIMES)) { return .... }
// rest of code
```

File: collector/processLink/helpers/downloadFileToHotDir.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
+const { WATCH_DIRECTORY } = require("../../utils/constants");
```
Comment (timothycarambat) [inline collector/processLink/helpers/downloadFileToHotDir.js]: Might make more sense to have this single util that downloads files moved to `collector/utils/files/index.js` since it has to do with files only and we can probably re-use it too

File: collector/processLink/helpers/downloadFileToHotDir.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+const fs = require("fs");
+const { pipeline } = require("stream/promises");
+/**
+ * Download a file to the hotdir
+ * @param {string} url - The URL of the file to download
+ */
+async function downloadFileToHotDir(url) {
```
Comment (timothycarambat) [inline collector/processLink/helpers/downloadFileToHotDir.js]: This is more aptly `downloadURItoFile` since it only takes a single URL that is `assumed` to be a file.

This function also does not do the check to validate if this file is downloadable, which could be confusing if used elsewhere in the future since it will just download any URL given (eg: `google.com`) and not specific ones that point to URL file-resources

Comment (timothycarambat) [reply collector/processLink/helpers/downloadFileToHotDir.js]: Might be a good opportunity to condense logic currently in `collector/processLink/convert/generic.js` to this function, which if doing so - this current helper file location would be more appropriate.

File: collector/processLink/helpers/downloadFileToHotDir.js [hunk 6]
```diff
@@ -0,0 +1,31 @@
+const { WATCH_DIRECTORY } = require("../../utils/constants");
+const fs = require("fs");
+const { pipeline } = require("stream/promises");
+/**
+ * Download a file to the hotdir
+ * @param {string} url - The URL of the file to download
```
Comment (timothycarambat) [inline collector/processLink/helpers/downloadFileToHotDir.js:9]: This JSDOC is incorrect, the function returns an object - not a string

File: collector/processLink/helpers/getContentType.js [hunk 7]
```diff
@@ -0,0 +1,25 @@
+async function getContentType(url) {
```
Comment (timothycarambat) [inline collector/processLink/helpers/getContentType.js]: Needs JSDoc

File: collector/processLink/helpers/getContentType.js [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+  // Determine the content type of the resource by making a HEAD request
+  const res = await fetch(url, { method: "HEAD" });
```
Comment (timothycarambat) [inline collector/processLink/helpers/getContentType.js]: This will throw an uncaught error if the passed in url is something invalid like `null` or undefined.
Utils, ideally, are atomic and self-contained and do their own validations within reason.

This allows us to also build unit tests for them in `jest` so we can test them for known edge cases.

File: collector/processLink/index.js [hunk 9]
```diff
@@ -13,7 +13,6 @@ async function processLink(link, scraperHeaders = {}, metadata = {}) {
   if (!validURL(link)) return { success: false, reason: "Not a valid URL." };
   return await scrapeGenericUrl({
     link,
-    captureAs: "text",
```
Comment (timothycarambat) [inline collector/processLink/index.js:16]: You will likely want to keep this. This specific execution of `scrapeGenericUrl` will always grab webpage text (since DOM is massive and messy for parsing and RAG). However, `scrapeGenericUrl` is also used below in `getLinkText` for agents/flows.

File: collector/utils/constants.js [hunk 10]
```diff
@@ -3,6 +3,10 @@ const WATCH_DIRECTORY = require("path").resolve(__dirname, "../hotdir");
 const ACCEPTED_MIMES = {
   "text/plain": [".txt", ".md", ".org", ".adoc", ".rst"],
   "text/html": [".html"],
+  "text/csv": [".csv"],
+  "application/json": [".json"],
+  // TODO: Create asDoc.js that works for standard MS Word files.
+  // "application/msword": [".doc"],
```
Comment (timothycarambat) [inline collector/utils/constants.js:9]: 👍

File: collector/utils/constants.js [hunk 11]
```diff
@@ -38,11 +42,16 @@ const SUPPORTED_FILETYPE_CONVERTERS = {
   ".org": "./convert/asTxt.js",
   ".adoc": "./convert/asTxt.js",
   ".rst": "./convert/asTxt.js",
+  ".csv": "./convert/asTxt.js",
+  ".json": "./convert/asTxt.js",
```
Comment (timothycarambat) [inline collector/utils/constants.js:47]: 👍

**Review by timothycarambat — CHANGES_REQUESTED** [review]

File: collector/processLink/convert/generic.js (hunk 1 above)
Comment (angelplusultra) [reply collector/processLink/convert/generic.js:22]: fix in [f882796](https://github.com/Mintplex-Labs/anything-llm/pull/4398/commits/f882796a157a9290e0bbb5d163dbd661664a92a6)

File: collector/processLink/convert/generic.js (hunk 2 above)
Comment (angelplusultra) [reply collector/processLink/convert/generic.js:27]: fix in https://github.com/Mintplex-Labs/anything-llm/pull/4398/commits/d033ad4953ba199767f1b25180223290ce2afdf2

File: collector/processLink/index.js (hunk 9 above)
Comment (angelplusultra) [reply collector/processLink/index.js:16]: fix in [f882796](https://github.com/Mintplex-Labs/anything-llm/pull/4398/commits/f882796a157a9290e0bbb5d163dbd661664a92a6)

File: collector/processLink/convert/generic.js (hunk 3 above)
Comment (angelplusultra) [reply collector/processLink/convert/generic.js]: fix in [2945c26](https://github.com/Mintplex-Labs/anything-llm/pull/4398/commits/2945c263ecf3edddbb9f13df8a764d6c1d19122a)

File: collector/processLink/helpers/getContentType.js (hunk 8 above)
Comment (angelplusultra) [reply collector/processLink/helpers/getContentType.js]: fix in [44f45b4](https://github.com/Mintplex-Labs/anything-llm/pull/4398/commits/44f45b47f29668e8f7920595330411345be5c686)

File: collector/processLink/helpers/getContentType.js (hunk 7 above)
Comment (angelplusultra) [reply collector/processLink/helpers/getContentType.js]: fix in [44f45b4](https://github.com/Mintplex-Labs/anything-llm/pull/4398/commits/44f45b47f29668e8f7920595330411345be5c686)

File: collector/processLink/helpers/downloadFileToHotDir.js (hunk 4 above)
Comment (angelplusultra) [reply collector/processLink/helpers/downloadFileToHotDir.js]: Fix in [0bba2c7](https://github.com/Mintplex-Labs/anything-llm/pull/4398/commits/0bba2c7aa5e3395cd592ab902ad25c088733638c)
